prometheus/inc/class-plugin.php

```diff
@@ -157,14 +92,71 @@ private static function create_registry(): RegistryInterface {
 		return new CollectorRegistry( $storage );
 	}
 
+	public function get_collectors(): array {
+		return $this->collectors;
+	}
+
 	/**
-	 * Validate if current request is for the Prometheus endpoint.
+	 * Cron callback to process the metrics for collection
 	 *
-	 * @return bool
+	 * @return void
+	 */
+	public function process_metrics(): void {
+		if ( Context::is_web_request() ) {
+			trigger_error( __METHOD__ . ' should not be called on web requests', E_USER_WARNING );
+			return;
+		}
+
+		foreach ( $this->collectors as $collector ) {
+			if ( is_callable( [ $collector, 'process_metrics' ] ) ) {
+				$collector->process_metrics();
+			}
+		}
+	}
+
+	/**
+	 * We're going to send the response to the client, then collect metrics from each registered collector
 	 */
-	private function is_prom_endpoint_request(): bool {
-		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized -- the value is used only for strict comparison
-		$request_uri = $_SERVER['REQUEST_URI'] ?? '';
-		return self::$endpoint_path === $request_uri;
+	public function shutdown(): void {
+		// This is expensive, potentially, so be mindful about when this method is called
+		// Currently it only runs on web requests with a 90 interval, see the constructor
+		if ( function_exists( 'fastcgi_finish_request' ) ) {
+			fastcgi_finish_request();
+		}
+
+		// Don't run on unauthorized requests.
+		if ( ! function_exists( 'wp_get_current_user' ) || ! current_user_can( 'edit_posts' ) ) {
+			return;
+		}
+
+		if ( wp_cache_add( 'vip_prometheus_last_collection_run', time(), 'vip_prometheus', 5 * MINUTE_IN_SECONDS ) ) {
+			foreach ( $this->collectors as $collector ) {
+				$collector->collect_metrics();
+			}
+
+			// Roughly every hour
+			if ( ! wp_next_scheduled( 'vip_prometheus_process_metrics' ) ) {
+				wp_schedule_single_event( time() + HOUR_IN_SECONDS, 'vip_prometheus_process_metrics' );
+			}
```

Comment:
I don't love adding in a new cron job for this 😬  - it particularly really starts to inflate cron jobs for large multisites. Would really like to start pairing up our internal jobs together on shared hooks.

Reply:
@WPprodigy  I don't love this either :)

Open to suggestions but I think here's what helps: the event is only scheduled when there's logged-in activity, so even on the large multisite the scenario where ALL bajillion sites are scheduling at the same time is of relatively low probability.

I'm open to suggestions though, alternative design would be to run on on a root site and do `switch_to_blog` but that may lead to some edge-cases or a long-running job that may be terminated by pod rolls.

Reply:
Addressed